README.md
<p align="center">
<img alt="Diary Worm Mascot" src="./img/diary-worm.png" width="250px" />
</p>

# Diary

This is a text-based diary, inspired by [khal](https://github.com/pimutils/khal). Journal entries are stored in text files.

## Demo
![Diary Demo](https://raw.githubusercontent.com/in0rdr/diary/master/img/demo.gif)

## Usage
1. Set the EDITOR environment variable to your favourite text editor:
    ```
    export EDITOR=vim
    ```

2. Run the diary, with the folder for the text files as first argument:
    ```
    diary ~/.diary
    ```

   Instead of this, you can also set the environment variable `DIARY_DIR`
   to the desired directory. If both an argument and the environment
   variable are given, the argument takes precedence (see [Variable Precedence Rules](#precedence_rules)).

   The text files in this folder will be named 'yyyy-mm-dd'.

3. Use the keypad or VIM-like shortcuts to move between dates:

    ```
    e, Enter  edit current entry
    d, x      delete current entry
    s         sync current entry with CalDAV server (ALPHA)
    S         sync all entries with CalDAV server (ALPHA)
    i         import entries from .ics file (ALPHA)

    t         jump to today
    f         jump to or find specific day

    j, down   go forward by 1 week
    k, up     go backward by 1 week
    h, left   go left by 1 day
    l, right  go right by 1 day
    J         go forward by 1 month
    K         go backward by 1 month

    N         go to the previous journal entry
    n         go to the next journal entry
    g         go to start of journal
    G         go to end of journal

    q         quit the program
    ```

![diary cheet sheet](https://raw.githubusercontent.com/in0rdr/diary/master/img/diary-cheat-sheet.png)

## Install

Pre-built packages can be installed from the [Open Build Service (OBS)](https://software.opensuse.org/download.html?project=home%3Ain0rdr&package=diary)
* Install the `diary` package for a stable releases
* Install the `diary-nightly` build to use a more recent but possibly unstable version from the master branch

### Arch Linux OBS Repository Bug

The Arch Linux repository on the OBS is broken, see [#69](https://github.com/in0rdr/diary/issues/69) and [#9953](https://github.com/openSUSE/open-build-service/issues/9953).

To use the repository on Arch Linux, use `downloadcontent.opensuse.org` instead of the [suggested](https://software.opensuse.org/download.html?project=home%3Ain0rdr&package=diary) name `download.opensuse.org` in the `/etc/pacman.conf`:
```
[home_in0rdr_Arch]
Server = https://downloadcontent.opensuse.org/repositories/home:/in0rdr/Arch/$arch
```

## Build from Source
[![Build Status](https://github.com/in0rdr/diary/actions/workflows/c.yml/badge.svg)](https://github.com/in0rdr/diary/actions/workflows/c.yml)

1. Define [OAuth2 application credentials](https://developers.google.com/identity/protocols/oauth2) if CalDAV sync should be effective:
    ```
    export GOOGLE_OAUTH_CLIENT_ID=""
    export GOOGLE_OAUTH_CLIENT_SECRET=""
    ```

    Alternatively, leave these two variables unset and [define the clientid/secret in the configuration file](#Google-Calendar-OAuth2) at run-time.

2. Compile (requires ncurses and libcurl):
    ```
    make
    ```
Note: for *BSD users run gmake.

3. Install the binary (optional):
    ```
    sudo make install
    ```

   By default this will copy the binary to /usr/local/bin. To use a different
   path prefix, type `sudo make PREFIX=/usr install` to use /usr/bin for example.
   You can uninstall diary with `sudo make uninstall`.

## OBS Build Documentation
[./docs/OBS.md](./docs/OBS.md)

## GitHub Action CI Documentation
[./docs/CI.md](./docs/CI.md)

## Test Documentation
[./docs/TESTING.md](./docs/TESTING.md)

## Configuration File

The [`diary.cfg`](./config/diary.cfg) configuration file can optionally be used to persist diary configuration. To install the sample from this repository:
```bash
mkdir -p ${XDG_CONFIG_HOME:-~/.config}/diary
cp config/diary.cfg ${XDG_CONFIG_HOME:-~/.config}/diary/
```

The file `${XDG_CONFIG_HOME:-~/.config}/diary/diary.cfg` should adhere to a basic `key = value` format. Lines can be commented with the special characters `#` or `;`. The following configuration keys are currently supported:

| Command Line Option | Config Key | Example Config Value | Default Config Value | Description |
| --- | --- | --- | --- | --- |
| `--dir`, `-d`, or first non-option argument | `dir` | ~/diary | n/a | Diary directory. Path that holds the journal text files. If unset, defaults to environment variable `$DIARY_DIR`.|
| `--editor` or `-e` | `editor` | vim | (empty) | Editor to open journal files with. If unset, defaults to environment variable `$EDITOR`. If no editor is provided, the diary is opened read-only. |
| `--fmt` or `-f` | `fmt` | %d_%b_%y | %Y-%m-%d | Date format and file name for the files inside the `dir`. For the format specifiers, see [`man strftime`](https://man7.org/linux/man-pages/man3/strftime.3.html). Be careful: If you change this, you might no longer find your existing diary entries, because the diary assumes to find the journal files under another file name. Hence, a change in FMT shows an empty diary, at first. Rename all files in the DIARY_DIR to migrate to a new FMT. |
| `--fmt-cmd` or `-F` | `fmt_cmd` | `fmt -w75 -s` | (empty) | Journal entries are formatted with this command. If not set (default), the characters are printed to the preview screen character by character (without word wrap). For example, to enable word wrap after 75 characters, use "fmt -s" as FMT_CMD. |
| `--range` or `-r` | `range` | 10 | 1 | Number of years to show before/after todays date |
| `--weekday` or `-w` | `weekday` | 0 | 1 | First weekday, `7` = Sunday, `1` = Monday, ..., `6` = Saturday. Use `7` (or `0`) to display week beginning at Sunday ("S-M-T-W-T-F-S"), or `1` for "M-T-W-T-F-S-S". If `glibc` is installed, the first day of the week is derived from the current locale setting (`$LANG`, see `man locale`). Without `glibc`, the first weekday defaults to 1 (Monday), unless specified otherwise with this option. |
| n/a | `google_calendar` | diary | (empty) | Displayname of Google Calendar for [CalDAV sync](#CalDAV-sync) |
| n/a | `google_clientid` | 123-123.apps.googleusercontent.com | [built-in](#Build) / (empty) | Google Calendar for [Google Calendar OAuth2](#Google-Calendar-OAuth2) clientid |
| n/a | `google_secretid` | 321 | [built-in](#Build) / (empty) |  Google Calendar for [Google Calendar OAuth2](#Google-Calendar-OAuth2) secretid |
| n/a | `google_tokenfile` | ~/.diary-token | ~/.diary-token | Displayname of Google Calendar for [Google Calendar OAuth2](#Google-Calendar-OAuth2) API token file|

## Precedence Rules
<a name="precedence_rules"></a>

The default variables, for instance, for the configuration variables `editor`, `dir` and `weekday`, are populated with values in the following order, where earlier entries are overwritten by subsequent ones if they exist:
1. No default for `DIARY_DIR`. Defaults for `range`, `weekday`, `fmt` and `editor` are provided in [diary.h](diary.h)
* If `EDITOR` is unset and no editor is provided in the config file or via the `-e` option, the diary works read-only. Journal files cannot be opened.
* If `DIARY_DIR` is not provided, the diary won't open.
2. **Config file** (empty default for `editor`, no default for `dir`)
3. **Environment** variables `$DIARY_DIR`, `$EDITOR` and `$LANG` for locale (`weekday`)
4. **Option arguments** `-d` / `-e` / `-w`
5. First non-option argument is interpreted as `DIARY_DIR`

### Precedence Example: Locale and First Day of Week

If glibc is installed, the first weekday defaults to the locale defined in the current shell environment (`$LANG`, see `man locale`), unless specified otherwise via the `--weekday` / `-w` command line options. For example:

```bash
# start with weekday=3(Wed), overrule any other configuration value
$ diary -w3

# start with glibc derived weekday=1, regardless of 'weekday' in config file
$ LANG=de_CH diary

# if glibc is installed, start with glibc derived base date (weekday=0)
$ LANG= diary

# disable environment variable, default to value from config file
$ unset LANG

# start with 'weekday' default from config file, if available
$ diary

# remove config file
$ rm ${XDG_CONFIG_HOME:-~/.config}/diary/diary.cfg

# start with 'weekday' default value from source code (1=Mon)
$ diary
```

## Import/Export (alpha)

The import functionality can be triggered by pressing `i`.

> ⚠️  Alpha feature: Available only in `diary-nightly` (see [Installation Instructions](#Install)). Don't use this with "production" data.

The export functionality is not implemented yet, see [issues/78](https://github.com/in0rdr/diary/issues/78).

## CalDAV Sync (alpha)
> ⚠️  Alpha feature: Available only in `diary-nightly` (see [Installation Instructions](#Install)). Don't use this with "production" data.

The journal files can be synced via CalDAV by pressing `s`. Currently, only the Google Calendar is supported as remote provider. Please open an [issue](https://github.com/in0rdr/diary/issues) to implement support for additional remote calendar servers.


The calender for synchronization can be defined with the [configuration](#Configuration-File) key `google_calendar`:
```
# Google calendar name for CalDAV sync
google_calendar = example
```

This key is empty by default and the only configuration key required for setting up synchronization.

### Google Calendar OAuth2

The Google Calendar [CalDAV API is protected with OAuth2](https://developers.google.com/identity/protocols/oauth2).

The credentials and the consent screen can be redefined at [compile time](#Build) or with the following keys in the [configuration file](#Configuration-File):

```
# Google OAuth2 clientid and secretid
google_clientid = 123-123.apps.googleusercontent.com
google_secretid = 321
```

The token used to authenticate with the Google API is stored in the file specified by `google_tokenfile` (defaults to `~/.diary-token`):
```
# Google OAuth2 tokenfile
google_tokenfile = ~/.diary-token
```

The application requires two [OAuth2 scopes](https://developers.google.com/calendar/auth) for CalDAV requests:

1. `https://www.googleapis.com/auth/calendar`: read/write access to Calendars - required to discover the unique hyperlink/URI for the calendar specified by the [configuration key](#Configuration-File) `google_calendar`
2. `https://www.googleapis.com/auth/calendar.events.owned`: read/write access to Events owned by the user - allows diary to create/read/update/delete events in `google_calendar`

## License

This work uses the MIT License. See [LICENSE](./LICENSE) for details.

## Contributors

All source code is available in this repository. Contributions are always welcome!

For a list of contributors see the [AUTHORS.md](./AUTHORS.md).

CHANGELOG.md
## [Unreleased]

### Changed
* New s shortcut logic: s no longer jumps to specific day (search). f shortcut is used instead to "find" date (drop-in replacement). s shortcut is used to "sync" entries to Google calendar (CalDAV sync).

### Added
* CalDAV sync: s/S shortcuts to sync entries to Google calendar.
* Import (#76, #77): i shortcut to import entries from .ics
* Export (#78): E shortcut to export entries to .ics
* External text format command (#68, #80) --fmt-cmd
* Mouse support for selecting dates (#60)
* Documentation for build pipelines (CI/CD) and Open Build Service (OBS) jobs (`docs/CI.md` and `docs/OBS.md`)
* Documentation of testing and debugging procedures (`docs/TESTING.md`)
* Add `debug` target to `Makefile`

## [0.4] - 2020-10-17

### Changed
* Fix install on OSX (#41)
* Improve Makefile and build instructions

## [0.3] - 2016-12-28

### Changed
* Escape characters and utf-8 fixes
* First day of week according to locale

### Added
* Header file
* Constant parameters
* Jump to nearest entry
* Man page

## [0.2] - 2016-11-27

### Changed
* Fixes trailing '/' in entry reading

### Added
* Remove entries
* Search entries, jump to date

## [0.1] - 2016-11-24

### Notes
* There is still a bug which appends '/' to the diary directory even if the trailing '/' is already given.
* Removing entries not yet possible

src/diary.h
#ifndef DIARY_H
#define DIARY_H

#ifdef __MACH__
	#include <CoreFoundation/CoreFoundation.h>
#endif
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#include <dirent.h>
#include <ncurses.h>
#include <locale.h>
#include <langinfo.h>
#include "utils.h"
#include "import.h"
#include "caldav.h"

#define XDG_CONFIG_HOME_FALLBACK "~/.config"
#define CONFIG_FILE_PATH "diary/diary.cfg"
#define DIARY_VERSION "0.6-unstable"

static const char* WEEKDAYS[] = {"Su","Mo","Tu","We","Th","Fr","Sa"};

void setup_cal_timeframe();
void draw_wdays(WINDOW* head);
void draw_calendar(WINDOW* number_pad, WINDOW* month_pad, const char* diary_dir, size_t diary_dir_size);

void display_entry(const char* dir, size_t dir_size, const struct tm* date, WINDOW* win, int width);
void edit_cmd(const char* dir, size_t dir_size, const struct tm* date, char** rcmd, size_t rcmd_size);

bool date_has_entry(const char* dir, size_t dir_size, const struct tm* i);

extern config CONFIG;

#endif

src/utils.h
#ifndef DIARY_UTILS_H
#define DIARY_UTILS_H

#include <regex.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <wordexp.h>
#include <stdbool.h>
#include <ncurses.h>

#define GOOGLE_OAUTH_TOKEN_FILE "~/.diary-token"
#ifndef GOOGLE_OAUTH_CLIENT_ID
    #define GOOGLE_OAUTH_CLIENT_ID ""
#endif
#ifndef GOOGLE_OAUTH_CLIENT_SECRET
    #define GOOGLE_OAUTH_CLIENT_SECRET ""
#endif

#define CAL_WIDTH 21
#define ASIDE_WIDTH 4
#define MAX_MONTH_HEIGHT 6

void update_date(WINDOW* header, struct tm* curs_date);
char* extract_json_value(const char* json, char* key, bool quoted);
char* fold(const char* str);
char* unfold(const char* str);
char* extract_ical_field(const char* ical, char* key, long* start_pos, bool multline);
char* expand_path(const char* str);
char* strrstr(char *haystack, char *needle);
void fpath(const char* dir, size_t dir_size, const struct tm* date, char** rpath, size_t rpath_size);
bool go_to(WINDOW* calendar, WINDOW* aside, time_t date, int* cur_pad_pos, struct tm* curs_date, struct tm* cal_start, struct tm* cal_end);
void* show_progress(void* vargp);

typedef struct
{
    // Path that holds the journal text files
    char* dir;
    // Number of years to show before/after todays date
    int range;
    // 7 = Sunday, 1 = Monday, ..., 6 = Saturday
    int weekday;
    // 2020-12-31
    char* fmt;
    // Text formatting command/utility fmt,
    // default width 75 chars (-w75), do not refill lines (-s)
    char* fmt_cmd;
    // Editor to open journal files with
    char* editor;
    // File for Google OAuth access token
    char* google_tokenfile;
    // Google client id
    char* google_clientid;
    // Google secret id
    char* google_secretid;
    // Google calendar to synchronize
    char* google_calendar;
} config;

extern config CONFIG;

#endif


src/caldav.h
#ifndef DIARY_CALDAV_H
#define DIARY_CALDAV_H

#define __USE_XOPEN
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/poll.h>
#include <netdb.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <ncurses.h>
#include <curl/curl.h>
#include "utils.h"

#define STR(s) #s
#define MKSTR(s) STR(s)

// https://developers.google.com/identity/protocols/oauth2/native-app#create-code-challenge
// A valid code_verifier has a length between 43 and 128 characters
#define GOOGLE_OAUTH_CODE_VERIFIER_LENGTH 43
#define GOOGLE_OAUTH_AUTHZ_URL "https://accounts.google.com/o/oauth2/auth"
#define GOOGLE_OAUTH_TOKEN_URL "https://oauth2.googleapis.com/token"
#define GOOGLE_OAUTH_SCOPE "https://www.googleapis.com/auth/calendar%20https://www.googleapis.com/auth/calendar.events.owned"
#define GOOGLE_OAUTH_RESPONSE_TYPE "code"
#define GOOGLE_OAUTH_REDIRECT_HOST "127.0.0.1"
#define GOOGLE_OAUTH_REDIRECT_PORT 9004
#define GOOGLE_OAUTH_REDIRECT_URI "http://" GOOGLE_OAUTH_REDIRECT_HOST ":" MKSTR(GOOGLE_OAUTH_REDIRECT_PORT)
#define GOOGLE_OAUTH_REDIRECT_SOCKET_BACKLOG 10
#define GOOGLE_API_URI "https://apidata.googleusercontent.com"
#define GOOGLE_CALDAV_URI GOOGLE_API_URI "/caldav/v2"

int caldav_sync(struct tm* date,
                 WINDOW* header,
                 WINDOW* cal,
                 int pad_pos,
                 const char* dir,
                 size_t dir_size,
                 bool confirm);

struct curl_mem_chunk {
    char* memory;
    size_t size;
};

extern config CONFIG;

#endif


src/import.c
#include "import.h"

/* Import journal entries from an ics file */
void ics_import(const char* ics_input, WINDOW* header, WINDOW* cal, WINDOW* aside, int* pad_pos, struct tm* curs_date, struct tm* cal_start, struct tm* cal_end) {
    pthread_t progress_tid;

    FILE* pfile = fopen(ics_input, "r");
    if (pfile == NULL) {
        perror("Error opening file");
        return;
    }

    fseek(pfile, 0, SEEK_END);
    long ics_bytes = ftell(pfile);
    rewind(pfile);

    char* ics = malloc(ics_bytes + 1);
    fread(ics, 1, ics_bytes, pfile);
    fclose(pfile);

    ics[ics_bytes] = '\0';
    // fprintf(stderr, "Import ICS file: %s\n", ics);

    int conf_ch = 0;
    char dstr[16];
    struct tm date = {};

    long search_pos = 0;
    char* vevent;
    char* vevent_date;
    char* vevent_desc;

    // find all VEVENTs and write to files
    char *i = ics;
    char* j;
    while (i < ics + ics_bytes) {
        vevent = extract_ical_field(i, "BEGIN:VEVENT", &search_pos, false);
        vevent_date = extract_ical_field(i, "DTSTART", &search_pos, false);
        vevent_desc = extract_ical_field(i, "DESCRIPTION", &search_pos, true);
        if (vevent == NULL || vevent_desc == NULL) {
            free(vevent);
            free(vevent_date);
            free(vevent_desc);
            break;
        }

        i += search_pos;

        // fprintf(stderr, "VEVENT DESCRIPTION: \n\n%s\n\n", vevent_desc);

        // parse date
        strptime(vevent_date, "%Y%m%d", &date);
        strftime(dstr, sizeof dstr, CONFIG.fmt, &date);

        // get path of entry
        char path[100];
        char* ppath = path;
        fpath(CONFIG.dir, strlen(CONFIG.dir), &date, &ppath, sizeof path);
        fprintf(stderr, "Import date file path: %s\n", path);

        if (conf_ch == 'c') {
            free(vevent);
            free(vevent_date);
            free(vevent_desc);
            break;
        }

        if (conf_ch != 'a') {
            // prepare header for confirmation dialogue
            curs_set(2);
            noecho();
            wclear(header);

            // ask for confirmation
            mvwprintw(header, 0, 0, "Import entry '%s' and overwrite local file? [(Y)es/(a)ll/(n)o/(c)ancel] ", dstr);
            conf_ch = wgetch(header);
        }

        if (conf_ch == 'y' || conf_ch == 'Y' || conf_ch == 'a' || conf_ch == '\n') {
            pthread_create(&progress_tid, NULL, show_progress, (void*)header);
            pthread_detach(progress_tid);

            // persist VEVENT to local file
            FILE* cursordate_file = fopen(path, "wb");
            if (cursordate_file == NULL) {
                perror("Failed to open import date file");
            } else {
                for (j = vevent_desc; *j != '\0'; j++) {
                    if (vevent_desc[j-vevent_desc] == 0x5C) { // backslash
                        switch (*(j+1)) {
                            case 'n':
                                fputc('\n', cursordate_file);
                                j++;
                                break;
                            case 0x5c: // preserve real backslash
                                fputc(0x5c, cursordate_file);
                                j++;
                                break;
                        }
                    } else {
                        fputc(*j, cursordate_file);
                    }
                }
                fclose(cursordate_file);

                bool mv_valid = go_to(cal, aside, mktime(&date), pad_pos, curs_date, cal_start, cal_end);
                if (mv_valid) {
                    // add new entry highlight
                    chtype atrs = winch(cal) & A_ATTRIBUTES;
                    wchgat(cal, 2, atrs | A_BOLD, 0, NULL);
                    prefresh(cal, *pad_pos, 0, 1, ASIDE_WIDTH, LINES - 1, ASIDE_WIDTH + CAL_WIDTH);
                }
            }
            pthread_cancel(progress_tid);
        }

        // fprintf(stderr, "Import DTSTART: %s\n", desc);
        // fprintf(stderr, "Import DESCRIPTION: %s\n", desc);
        fprintf(stderr, "* * * * * * * * * * * * * \n");
        free(vevent);
        free(vevent_date);
        free(vevent_desc);
    }
    free(ics);
}

src/utils.c
#include "utils.h"

/* Update the header with the cursor date */
void update_date(WINDOW* header, struct tm* curs_date) {
    // TODO: dstr for strlen(CONFIG.format) > 16 ?
    char dstr[16];
    mktime(curs_date);
    strftime(dstr, sizeof dstr, CONFIG.fmt, curs_date);

    wclear(header);
    mvwaddstr(header, 0, 0, dstr);
    wrefresh(header);
}


char* extract_json_value(const char* json, char* key, bool quoted) {
    // work on a copy of the json
    char* jsoncp = (char*) malloc(strlen(json) * sizeof(char) + 1);
    if (jsoncp == NULL) {
        perror("malloc failed");
        return NULL;
    }
    strcpy(jsoncp, json);

    char* tok = strtok(jsoncp, " ");
    while (tok != NULL) {
        if (strstr(tok, key) != NULL) {
            tok = strtok(NULL, " "); // value
            break;
        }
        // key was not in this tok, advance tok
        tok = strtok(NULL, " ");
    }

    // remove quotes and comma or commma only
    if (quoted) {
        tok = strtok(tok, "\"");
    } else {
        tok = strtok(tok, ",");
    }

    char* res = NULL;
    if (tok != NULL) {
        res = (char*) malloc(strlen(tok) * sizeof(char) + 1);
        if (res == NULL) {
            perror("malloc failed");
            return NULL;
        }
        strcpy(res, tok);
    }

    free(jsoncp);
    return res;
}

char* fold(const char* str) {
    // work on a copy of the str
    char* strcp = (char *) malloc(strlen(str) * sizeof(char) + 1);
    if (strcp == NULL) {
        perror("malloc failed");
        return NULL;
    }
    strcpy(strcp, str);

    // create buffer for escaped result TEXT
    char* buf = malloc(1);
    if (buf == NULL) {
        perror("malloc failed");
        return NULL;
    }
    buf[0] = '\0';

    void* newbuf;
    // bufl is the current buffer size incl. \0
    int bufl = 1;
    // i is the iterator in strcp
    char* i = strcp;
    // escch is the char to be escaped,
    // only written when esc=true
    char escch;
    bool esc = false;

    while(*i != '\0' || esc) {
        fprintf(stderr, "strlen(buf): %i\n", bufl);
        fprintf(stderr, "*i: %c\n", *i);
        fprintf(stderr, "escch: %c\n", escch);
        fprintf(stderr, "esc: %i\n", esc);
        fprintf(stderr, "buffer: %s\n\n", buf);

        newbuf = realloc(buf, ++bufl);
        if (newbuf == NULL) {
            perror("realloc failed");
            free(buf);
            return NULL;
        }
        buf = (char*) newbuf;

        if ((bufl > 1) && ((bufl % 77) == 0)) {
            // break lines after 75 chars
            // split between any two characters by inserting a CRLF
            // immediately followed by a white space character
            buf[bufl-2] = '\n';
            escch = ' ';
            esc = true;
            continue;
        }

        if (esc) {
            // only escape char, do not advance iterator i
            buf[bufl-2] = escch;
            esc = false;
        } else {
            // escape characters
            // https://datatracker.ietf.org/doc/html/rfc5545#section-3.3.11
            switch (*i) {
                case 0x5c: // backslash
                    buf[bufl-2] = 0x5c;
                    escch = 0x5c;
                    esc = true;
                    break;
                case ';':
                    buf[bufl-2] = 0x5c;
                    escch = ';';
                    esc = true;
                    break;
                case ',':
                    buf[bufl-2] = 0x5c;
                    escch = ',';
                    esc = true;
                    break;
                case '\n':
                    buf[bufl-2] = 0x5c;
                    escch = 'n';
                    esc = true;
                    break;
                default:
                    // write regular character from strcp
                    buf[bufl-2] = *i;
                    break;
            }
            i++;
        }

        // terminate the char string in any case (esc or not)
        buf[bufl-1] = '\0';
    }

    fprintf(stderr, "escch: %c\n", escch);
    fprintf(stderr, "end: %c\n", buf[bufl]);

    free(strcp);
    return buf;
}

char* unfold(const char* str) {
    fprintf(stderr, "Before unfolding: %s\n", str);

    // work on a copy of the str
    char* strcp = (char *) malloc(strlen(str) * sizeof(char) + 1);
    if (strcp == NULL) {
        perror("malloc failed");
        return NULL;
    }
    strcpy(strcp, str);

    char* res = strtok(strcp, "\n");

    if (res == NULL) {
        fprintf(stderr, "No more lines in multiline string, stop unfolding.\n");
        free(strcp);
        return NULL;
    }

    char* buf = malloc(strlen(res) + 1);
    if (buf == NULL) {
        perror("malloc failed");
        return NULL;
    }
    strcpy(buf, res);

    char* newbuf;
    regex_t re;
    regmatch_t pm[1];

    if (regcomp(&re, "^[^ \t]", 0) != 0) {
        perror("Failed to compile regex");
        return NULL;
    }

    while (res != NULL) {
        res = strtok(NULL, "\n");

        if (res == NULL) {
            fprintf(stderr, "No more lines in multiline string, stop unfolding.\n");
            break;
        }

        if (regexec(&re, res, 1, pm, 0) == 0) {
            // Stop unfolding if line does not start with white space/tab:
            // https://datatracker.ietf.org/doc/html/rfc2445#section-4.1
            break;
        }

        newbuf = realloc(buf, strlen(buf) + strlen(res) + 1);
        if (newbuf == NULL) {
            perror("realloc failed");
            break;
        } else {
            buf = newbuf;
            strcat(buf, res + 1);
        }
    }

    regfree(&re);
    free(strcp);
    return buf;
}

/* Find ical key in string. The value of 'start_pos' is set to the start position
   of the value (match) after the colon (':').
*/
char* extract_ical_field(const char* ics, char* key, long* start_pos, bool multiline) {
    regex_t re;
    regmatch_t pm[1];
    char key_regex[strlen(key) + 1];
    sprintf(key_regex, "^%s", key);
    fprintf(stderr, "Key regex: %s\n", key_regex);

    if (regcomp(&re, key_regex, 0) != 0) {
        perror("Failed to compile regex");
        return NULL;
    }

    // work on a copy of the ical xml response
    char* icscp = (char *) malloc(strlen(ics) + 1 * sizeof(char));
    if (icscp == NULL) {
        perror("malloc failed");
        return NULL;
    }
    strcpy(icscp, ics);

    // tokenize ical by newlines
    char* buf = NULL;
    char* res = strtok(icscp, "\n");
    while (res != NULL) {
        if (regexec(&re, res, 1, pm, 0) == 0) {
            // found the key in line 'res'
            res = strstr(res, ":"); // value
            res++; // strip the ":"

            fprintf(stderr, "Extracted ical result value: %s\n", res);
            fprintf(stderr, "Extracted ical result size: %li\n", strlen(res));

            fprintf(stderr, "Sizeof ics: %li\n", strlen(ics));
            fprintf(stderr, "Start pos: %li\n", *start_pos);
            fprintf(stderr, "Res: %s\n", res);
            *start_pos = res - icscp;
            fprintf(stderr, "Start pos: %li\n", *start_pos);

            if (strlen(res) != 0) {
                // non-empty remote value
                if (multiline) {
                    buf = unfold(ics + *start_pos);
                } else {
                    buf = malloc(strlen(res) + 1);
                    if (buf == NULL) {
                        perror("malloc failed");
                        return NULL;
                    }
                    strcpy(buf, res);
                }
            }
            break;
        }
        res = strtok(NULL, "\n");
    }

    regfree(&re);
    free(icscp);
    return buf;
}

// Return expanded file path
char* expand_path(const char* str) {
    char* res = "";
    wordexp_t str_wordexp;
    if ( wordexp( str, &str_wordexp, 0 ) == 0) {
        res = (char *) calloc(strlen(str_wordexp.we_wordv[0]) + 1, sizeof(char));
        strcpy(res, str_wordexp.we_wordv[0]);
    }
    wordfree(&str_wordexp);
    return res;
}

// Get last occurence of string in string
// https://stackoverflow.com/questions/20213799/finding-last-occurence-of-string
char* strrstr(char *haystack, char *needle) {
    char *i;
    int nlen = strlen(needle);
    for (i = haystack + strlen(haystack) - nlen; i >= haystack; i--) {
        if (strncmp(i, needle, nlen) == 0) {
            return i;
        }
    }
    return NULL;
}

/* Writes file path for 'date' entry to 'rpath'. '*rpath' is NULL on error. */
void fpath(const char* dir, size_t dir_size, const struct tm* date, char** rpath, size_t rpath_size)
{
    // check size of result path
    if (dir_size + 1 > rpath_size) {
        fprintf(stderr, "Directory path too long");
        *rpath = NULL;
        return;
    }

    // add path of the diary dir to result path
    strcpy(*rpath, dir);

    // check for terminating '/' in path
    if (dir[dir_size - 1] != '/') {
        // check size again to accommodate '/'
        if (dir_size + 1 > rpath_size) {
            fprintf(stderr, "Directory path too long");
            *rpath = NULL;
            return;
        }
        strcat(*rpath, "/");
    }

    char dstr[16];
    strftime(dstr, sizeof dstr, CONFIG.fmt, date);

    // append date to the result path
    if (strlen(*rpath) + strlen(dstr) > rpath_size) {
        fprintf(stderr, "File path too long");
        *rpath = NULL;
        return;
    }
    strcat(*rpath, dstr);
}

bool go_to(WINDOW* calendar, WINDOW* aside, time_t date, int* cur_pad_pos, struct tm* curs_date, struct tm* cal_start, struct tm* cal_end) {
    if (date < mktime(cal_start) || date > mktime(cal_end)) {
        return false;
    }

    int diff_seconds = date - mktime(cal_start);
    int diff_days = diff_seconds / 60 / 60 / 24;
    int diff_weeks = diff_days / 7;
    int diff_wdays = diff_days % 7;

    localtime_r(&date, curs_date);

    int cy, cx;
    getyx(calendar, cy, cx);

    // remove the STANDOUT attribute from the day we are leaving
    chtype current_attrs = mvwinch(calendar, cy, cx) & A_ATTRIBUTES;
    // leave every attr as is, but turn off STANDOUT
    current_attrs &= ~A_STANDOUT;
    mvwchgat(calendar, cy, cx, 2, current_attrs, 0, NULL);

    // add the STANDOUT attribute to the day we are entering
    chtype new_attrs =  mvwinch(calendar, diff_weeks, diff_wdays * 3) & A_ATTRIBUTES;
    new_attrs |= A_STANDOUT;
    mvwchgat(calendar, diff_weeks, diff_wdays * 3, 2, new_attrs, 0, NULL);

    if (diff_weeks < *cur_pad_pos)
        *cur_pad_pos = diff_weeks;
    if (diff_weeks > *cur_pad_pos + LINES - 2)
        *cur_pad_pos = diff_weeks - LINES + 2;
    prefresh(aside, *cur_pad_pos, 0, 1, 0, LINES - 1, ASIDE_WIDTH);
    prefresh(calendar, *cur_pad_pos, 0, 1, ASIDE_WIDTH, LINES - 1, ASIDE_WIDTH + CAL_WIDTH);

    return true;
}


void* show_progress(void* vargp){
    WINDOW* header = (WINDOW*) vargp;
    mvwprintw(header, 0, COLS - CAL_WIDTH - ASIDE_WIDTH - 11, "   syncing ");
    for(;;) {
        mvwprintw(header, 0, COLS - CAL_WIDTH - ASIDE_WIDTH - 10, "|");
        wrefresh(header);
        usleep(200000);
        mvwprintw(header, 0, COLS - CAL_WIDTH - ASIDE_WIDTH - 10, "/");
        wrefresh(header);
        usleep(200000);
        mvwprintw(header, 0, COLS - CAL_WIDTH - ASIDE_WIDTH - 10, "-");
        wrefresh(header);
        usleep(200000);
        mvwprintw(header, 0, COLS - CAL_WIDTH - ASIDE_WIDTH - 10, "\\");
        wrefresh(header);
        usleep(200000);
    }
}

config CONFIG = {
    .dir = NULL,
    .range = 1,
    .weekday = 1,
    .fmt = "%Y-%m-%d",
    .fmt_cmd = "",
    .editor = "",
    .google_tokenfile = GOOGLE_OAUTH_TOKEN_FILE,
    .google_clientid = GOOGLE_OAUTH_CLIENT_ID,
    .google_secretid = GOOGLE_OAUTH_CLIENT_SECRET,
    .google_calendar = ""
};
